### ISSUE TITLE
nextjs app deployed to vercel edge can't import prisma WASM module

### ISSUE THREAD

**Issue by zhelezkov** [prisma/prisma#24673]
### Bug description

Hey all,

I recently created a new nextjs project from scratch and was planning to use prisma as always, however I faced an issue using it with vercel edge environment. This issue is valid only for prisma 5.16.x, because 5.15.x works well. I suspect this commit caused this issue https://github.com/prisma/prisma/commit/63793a19799e4b181d76aa06863de47983ac3763

Seems like a nextjs bundler can't import wasm module with default configuration. Please see steps to reproduce for more details and error log

### How to reproduce

1. Create a new nextjs project with `npx create-next-app@latest`
2. Configure prisma according to official guide https://www.prisma.io/docs/orm/prisma-client/deployment/edge/deploy-to-vercel
3. add `export const runtime = "edge"` to your nextjs app, so it'll use an edge environment
4. Run it locally or deploy to vercel
5. See error
```
ModuleParseError: Module parse failed: Unexpected character '' (1:0)
The module seem to be a WebAssembly module, but module is not flagged as WebAssembly module for webpack.
BREAKING CHANGE: Since webpack 5 WebAssembly is not enabled by default and flagged as experimental feature.
You need to enable one of the WebAssembly experiments via 'experiments.asyncWebAssembly: true' (based on async modules) or 'experiments.syncWebAssembly: true' (like webpack 4, deprecated).
For files that transpile to WebAssembly, make sure to set the module type in the 'module.rules' section of the config (e. g. 'type: "webassembly/async"').
(Source code omitted for this binary file)
```

Enabling `experiments.syncWebAssembly` in `next.config.mjs` fixes a problem
```next.config.mjs
/** @type {import('next').NextConfig} */
const nextConfig = {
  webpack: (config, context) => {
    // console.log(config)
    config.experiments = {
      ...config.experiments,
      asyncWebAssembly: true,
    };
    return config;
  },
};

export default nextConfig;
```

### Expected behavior

I expected such setup to be working according to doc as it was in prisma 5.15.x

### Prisma information

no specific prisma.schema or prisma query is needed, since it's a build time issue, but I have `driverAdapters` feature enabled
```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["driverAdapters"]
}
```

### Environment & setup

- OS: MacOS and Vercel Edge environment
- Database: Neon/PostgreSQL
- Node.js version: v20.13.1

### Prisma Version

```
Environment variables loaded from .env
prisma                  : 5.16.1
@prisma/client          : 5.16.1
Computed binaryTarget   : darwin-arm64
Operating System        : darwin
Architecture            : arm64
Node.js                 : v20.13.1
Query Engine (Node-API) : libquery-engine 34ace0eb2704183d2c05b60b52fba5c43c13f303 (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli 34ace0eb2704183d2c05b60b52fba5c43c13f303 (at node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.16.0-24.34ace0eb2704183d2c05b60b52fba5c43c13f303
Default Engines Hash    : 34ace0eb2704183d2c05b60b52fba5c43c13f303
Studio                  : 0.502.0
Preview Features        : driverAdapters
```

**Comment by haydenbleasel** [comment]
Downgrading to `5.15.1` fixes this in the meantime 👍

**Comment by zhelezkov** [comment]
> Downgrading to `5.15.0` fixes this in the meantime 👍

yep, also I described another workaround which fixes it as well(adding a `experiments.asyncWebAssembly` in webpack config)

**Comment by ShamanAndrey** [comment]
Can **confirm** downgrading back to `5.15.0` works

**Comment by jkomyno** [comment]
Hi, I investigated this issue, here are some updates:
- This regression was introduced in https://github.com/prisma/prisma/pull/24554, which fixed a tooling compatibility problem with Vitest (which doesn't support dynamic conditional imports out of the box)
- It caused some breaks in Next.js with the Edge runtime (which e.g. is enabled by default when using middlewares) due to `webpack`. In fact, `webpack` statically analyses all imports at `next build` time, even the ones that would never occur at runtime (like [this one](https://github.com/prisma/prisma/blob/d433bb71df3bc7e2f9bdfba298177bfeebdf5d7b/packages/client/src/generation/utils/buildGetQueryEngineWasmModule.ts#L45)).

A possible fix to this problem would be adding the `/* webpackIgnore */` [magic comment](https://webpack.js.org/api/module-methods/#magic-comments), but even that doesn't seem to work, as tested by manually patching `@prisma/client` in [this example project](https://github.com/prisma/ecosystem-tests/tree/0da8268f40fe9cde9998a01cc39c250b313a7d53/driver-adapters-wasm/turso-vercel-nextjs-edgemw). Looks like `webpack` is ignoring these comment directives when it come to `.wasm` files.

An alternative course of action would be to revert https://github.com/prisma/prisma/pull/24554, but that would break `prisma` + `vitest` support again (possibly impacting a smaller audience, but still another possible regression nonetheless).

As of now, it seems that the WebAssembly support is still quite fragmented, and different behaviors emerge depending on the underlying bundler (e.g., `webpack`, `vite`, `esbuild`, etc). We're continuing this investigation internally, and we'll keep you promptly updated.

**Comment by jkomyno** [comment]
Some updates: I've released an unstable integration version, `@prisma/client@5.17.0-integration-fix-wasm-nextjs.1`.
It seems that, although it works locally with Next.js - while also not breaking Vitest support -, it still fails when deployed on Vercel ([see comment](https://github.com/prisma/prisma/pull/24711#issuecomment-2210953673)) :/

This work is still in progress.

**Comment by SevInf** [comment]
Yet another update: I think we finally have a working all the way throught fix for this issue that does not break `vitest` as well. @zhelezkov @haydenbleasel @ShamanAndrey could any of you try updating all prisma packages to `5.17.0-integration-fix-vitest-edgemw.1` version and see if that fixes the issue for you? We don't recommend using integration versions in production, but it should be good enough to verify the fix.

**Comment by zhelezkov** [comment]
Hey, I tested and version `5.17.0-integration-fix-vitest-edgemw.1` works great for me on vercel edge environment.
Thanks a lot! Looking forward to 5.17.0 release with this fix included 🙂

**Comment by SevInf** [comment]
Thanks for confirming the fix, no need to wait until 5.17.0 - we've just released it as 5.16.2.
